### PR TITLE
METRON-1052: Add forensic similarity hash functions to Stellar

### PR DESCRIPTION
**Pull request by cestella**
## Contributor Comments
This is a follow-on to METRON-539. Currently we have Stellar functions to perform cryptographic hashing operations. It would be useful to expand this to support forensic similarity hash functions so we could compare the similarity of inputs.  I have added support in this PR for [TLSH](https://github.com/trendmicro/tlsh/blob/master/TLSH_CTC_final.pdf)

This fits well within the existing `HASH` abstractions.  I have expanded and generalized it in a few places, but I think it's all within the spirit of the thing.

I owe a use-case driven walk-through of a demo that I designed, but haven't written up yet.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && build_utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/utils/hashing/TLSHHasher.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+   * this implementation does throw exceptions when the input is null.
+   *
+   * @param o The value to hash.
+   * @return A hash of {@code toHash} that has been encoded.
+   * @throws EncoderException         If unable to encode the hash then this exception occurs.
+   * @throws NoSuchAlgorithmException If the supplied algorithm is not known.
+   */
+  @Override
+  public Object getHash(Object o) throws EncoderException, NoSuchAlgorithmException {
+    TlshCreator creator = new TlshCreator(bucketOption, checksumOption);
```
Comment (ottobackwards) [inline metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/utils/hashing/TLSHHasher.java]: Can these be cached and reused?  Similar to how we cache regex patterns?

File: metron-stellar/stellar-common/pom.xml [hunk 2]
```diff
@@ -51,6 +52,11 @@
             </exclusions>
         </dependency>
         <dependency>
```
Comment (ottobackwards) [inline metron-stellar/stellar-common/pom.xml]: Does there need to be a dependencies csv entry for this?

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/utils/hashing/TLSHHasher.java (hunk 1 above)
Comment (cestella) [reply metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/utils/hashing/TLSHHasher.java]: yeah, actually, that's a damned fine suggestion.

File: metron-stellar/stellar-common/pom.xml (hunk 2 above)
Comment (cestella) [reply metron-stellar/stellar-common/pom.xml]: Yep, added and dealt with merge conflicts.

**Comment by ottobackwards** [comment]
+1 by inspection, really nice.

File: metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/utils/hashing/TLSHHasher.java (hunk 1 above)
Comment (cestella) [reply metron-stellar/stellar-common/src/main/java/org/apache/metron/stellar/common/utils/hashing/TLSHHasher.java]: Ok, I set up a cache to create and reuse the creators.  I also added a test case ensuring that the functions are all threadsafe with the use of the cache.

**Comment by cestella** [comment]
Let me write up a step-by-step use-case doc and I'll call this one done.

**Comment by ottobackwards** [comment]
Ok, I have not worked through the doc, but it looks good to me.  I can stop thinking about how this relates to FUZZY_SCORE.

**Comment by mattf-horton** [comment]
FUZZY_SCORE and BLOOM_*.  Is it worth trying to unify these similarity-recognizers with similarity hash, or are they too far apart in terms of expected use patterns?

**Comment by mattf-horton** [comment]
Altho I suppose the "Locality-Sensitive" part of TLSH means it operates at word level instead of byte or character level?

**Comment by ottobackwards** [comment]
One day we will have a handle book, the Tao of Stellar, that will have use cases to function mapping and examples

**Pull request closed without merging**